src/ui/textfield.rs

```diff
@@ -85,56 +101,199 @@ impl TextField {
             canvas.draw_rrect(rrect, &paint);
         });
 
-        // rendering: text
+        // Rendering: text
         ui.push_group(ui.size(), Layout::Freeform);
         ui.pad((16.0, 0.0));
         canvas.save();
         ui.clip(canvas);
 
-        // render hint
+        // Rendering: hint
         if hint.is_some() && self.text.len() == 0 {
             ui.text(canvas, hint.unwrap(), colors.text_hint, (AlignH::Left, AlignV::Middle));
         }
+
+        ui.text(canvas, &self.text_utf8, colors.text, (AlignH::Left, AlignV::Middle));
+
         let text_advance = ui.text(canvas, &self.text_utf8, colors.text, (AlignH::Left, AlignV::Middle));
 
         if self.focused && (input.time_in_seconds() - self.blink_start) % Self::BLINK_PERIOD < Self::HALF_BLINK {
             ui.draw_on_canvas(canvas, |canvas| {
                 let mut paint = Paint::new(Color4f::from(colors.text), None);
                 paint.set_anti_alias(false);
                 paint.set_style(paint::Style::Stroke);
-                let x = text_advance + 1.0;
+
+                let curr_text: String = self.text[..self.cursor_pos].iter().collect();
+                let cursor_x = ui.borrow_font().measure_str(curr_text, None).0;
+
+                let x = cursor_x + 1.0;
                 let y1 = Self::height(ui) * 0.2;
                 let y2 = Self::height(ui) * 0.8;
                 canvas.draw_line((x, y1), (x, y2), &paint);
             });
         }
 
+        if self.cursor_pos != self.selection_cursor_pos {
+            ui.draw_on_canvas(canvas, |canvas| {
+                let mut paint = Paint::new(Color4f::from(colors.selection), None);
+                paint.set_anti_alias(true);
+
+                let pre_selection_text: String = self.text[..self.selection_cursor_pos].iter().collect();
```

Comment:
As far as I could tell, this system of yours assumes that the selection is always made from left to right, which isn't always the case. A better solution would be to store `start` and `end` cursors. It makes a lot of the math easier and would probably work better with all the string slicing shenanigans that are needed to obtain `&str`s from the text field's text (see the `text_utf8` field of `struct TextField`).